### PR TITLE
Kotlin/Common Buffer read and write hexadecimal unsigned long

### PR DESCRIPTION
**Pull request by bnorm**
Found a couple algorithms on http://aggregate.org/MAGIC which were very helpful in replacing the Java specific methods for calculating the width required for a hex unsigned long. I'm hoping these are safe to use? I've linked the specific algorithm in the code.

### CODE REVIEW

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt [hunk 1]
```diff
@@ -565,6 +608,124 @@ internal inline fun Buffer.commonRead(sink: ByteArray, offset: Int, byteCount: I
...
+        if (seen == 0) {
+          throw NumberFormatException(
+            "Expected leading [0-9a-fA-F] character but was ${b.toHexString()}")
+        }
+        // Set a flag to stop iteration. We still need to run through segment updating below.
+        done = true
+        break
+      }
+
+      // Detect when the shift will overflow.
```
Comment (swankjesse) [inline okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: interesting, both  this and the decimal case support an arbitrary number of leading `0` bytes

Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: I’m sure I already knew that, but now I’m relearning it!

**Comment by swankjesse** [comment]
Sorry for sitting on this. Great PR.

File: okio/jvm/jmh/src/jmh/java/com/squareup/okio/benchmarks/WriteHexadecimalBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+WriteHexadecimalBenchmark.writeHex       11  thrpt    5  24.149 ± 0.943  ops/us
+WriteHexadecimalBenchmark.writeHex       12  thrpt    5  23.532 ± 2.501  ops/us
+WriteHexadecimalBenchmark.writeHex       13  thrpt    5  24.034 ± 0.166  ops/us
+WriteHexadecimalBenchmark.writeHex       14  thrpt    5  24.080 ± 0.388  ops/us
+WriteHexadecimalBenchmark.writeHex       15  thrpt    5  23.979 ± 0.414  ops/us
+WriteHexadecimalBenchmark.writeHex       16  thrpt    5  23.535 ± 0.116  ops/us
+ * </pre>
+ *
+ * Current performance:<pre>
+Benchmark                           (width)   Mode  Cnt   Score   Error   Units
```
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/jvm/jmh/src/jmh/java/com/squareup/okio/benchmarks/WriteHexadecimalBenchmark.java]: This is only the summary of the results so I believe it is averaging the 5 runs it performs for each parameter combination.

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt [hunk 3]
```diff
@@ -474,6 +474,49 @@ internal inline fun Buffer.commonWriteDecimalLong(v: Long): Buffer {
   return this
 }
 
+internal inline fun Buffer.commonWriteHexadecimalUnsignedLong(v: Long): Buffer {
+  var v = v
+  if (v == 0L) {
+    // Both a shortcut and required since the following code can't handle zero.
```
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: This one is copied from the original function so right back at you. 😉

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt [hunk 4]
```diff
@@ -474,6 +474,49 @@ internal inline fun Buffer.commonWriteDecimalLong(v: Long): Buffer {
...
+  // http://aggregate.org/MAGIC/#Population%20Count%20(Ones%20Count)
+  x -= x ushr 1 and 0x5555555555555555
+  x = (x ushr 2 and 0x3333333333333333) + (x and 0x3333333333333333)
+  x = (x ushr 4) + x and 0x0f0f0f0f0f0f0f0f
+  x += x ushr 8
+  x += x ushr 16
+  x = (x and 0x3f) + ((x ushr 32) and 0x3f)
+
+  // Round up to the nearest full byte
+  val width = ((x + 3) / 4).toInt()
```
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: I'll check and update accordingly.

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt [hunk 5]
```diff
@@ -565,6 +608,124 @@ internal inline fun Buffer.commonRead(sink: ByteArray, offset: Int, byteCount: I
...
+internal inline fun Buffer.commonReadDecimalLong(): Long {
+  if (size == 0L) throw EOFException()
+
+  // This value is always built negatively in order to accommodate Long.MIN_VALUE.
+  var value = 0L
+  var seen = 0
+  var negative = false
+  var done = false
+
+  val overflowZone = Long.MIN_VALUE / 10L
```
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: I will pull this and the following out as `const`.

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt [hunk 6]
```diff
@@ -565,6 +608,124 @@ internal inline fun Buffer.commonRead(sink: ByteArray, offset: Int, byteCount: I
...
+    if (pos == limit) {
+      head = segment.pop()
+      SegmentPool.recycle(segment)
+    } else {
+      segment.pos = pos
+    }
+  } while (!done && head != null)
+
+  size -= seen.toLong()
+  return if (negative) value else -value
```
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: Other than the `toHexString()` change for the error path, these read methods were copied without modification but I'll check to make sure we have that case unit tested.

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: Benchmark didn't show any change so I'll leave as is unless you prefer the shift from a style perspective: `((x + 3) shr 2).toInt()`

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnorm) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: Good catch: #628

File: okio/src/commonMain/kotlin/okio/internal/Buffer.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (swankjesse) [reply okio/src/commonMain/kotlin/okio/internal/Buffer.kt]: Keep it as-is!

**Pull request merged**